018-function-execution-model-refactor.md

```diff
@@ -0,0 +1,89 @@
+# Function Execution Model Rewrite
+- Implementation Owner: @PineappleIOnic
+- Start Date: 25-07-2021
+- Target Date: TBD
+
+## Summary
+
+[summary]: #summary
+
+<!-- Brief explanation of the proposed contribution. Write your answer below. -->
+This RFC contains a proposal for rewriting Cloud Functions in a way that will enable us to perform synchronous executions. This will introduce a new server into Appwrite specifically for dealing with executions called the "executor". The executor will deal with spinning up docker containers for runtimes that are required as well as updating containers with new code when needed. When API requests for functions are received the executor will forward them to the relevant runtime which will then handle it.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Cloud Functions currently do not support the ability to execute synchronous functions and will require a complete rethink to how we approach how we execute our functions in order for us to realize this goal.
+
+**What is the context or background in which this problem exists?**
+
+Cloud Functions are currently a bit limited, this new design idea will allow for more customizability and control of how functions will execute and run for users.
+
+**Once the proposal is implemented, how will the system change?**
+
+- A new process will be introduced to execute functions
+- All current runtimes will be updated to use a web server as it's core
+- The Functions worker will be stripped of all direct docker interactions and will instead call the executor
```

Comment:
If the Functions worker is stripped of all its direct Docker work, what will be left?

Reply:
Scheduled and async executions?

Reply:
Looking at the Architecture diagram, there's no significant workflow difference between sync and async functions - in fact, the Functions worker doesn't appear to have any function at all if we follow that document.

With so few responsibilities, I think it's confusing to have both a Functions worker (responsible for very little) alongside the Executor, which does basically everything else. Maybe there's a more appropriate name for the Functions worker after this change?